### DISCUSSION TITLE
Fritzbox self-signed SSL Certificate

### DISCUSSION THREAD

**Discussion by Robi1010MD** [mikejac/node-red-contrib-google-smarthome#424] [Q&A]
Hi,

I am currently struggling with the fact that although I followed the instructions, have a domain and an SSL certificate, I cannot connect. Port forwarding is also enabled. 

The Fritzbox uses port 3000 while the device uses port 3001. 

I got the Error: ERR_CONNECTION_REFUSED

So what could be the reason that I can't connect to the device?

**Comment by Caprico85** [comment]
Hello,

What message do you get and where do you get the message? When you try to open https://example.com:3000/check (with your domain) in the browser? Or during another step of the setup process?

**Comment by Robi1010MD** [comment]
Hey,thx for your fast reply. I got this Message, when i try to Open my specific Url and Check if it works.

**Comment by Caprico85** [comment]
Okay. Let's go through this step by step.

The port forwarding settings on the Fritzbox should look like this:

![image](https://github.com/mikejac/node-red-contrib-google-smarthome/assets/2081806/f567b6a1-04b1-4230-8610-891e6924dde6)

The web server settings in the configuration of the management node should look like this:

![image](https://github.com/mikejac/node-red-contrib-google-smarthome/assets/2081806/fb9128d5-0dbd-494a-9637-d3a3974972de)
(with other paths to the certificate)

If you use Docker, port forwardings there must be configured too.
If you use a reverse proxy, that must be configured as well.

If you are running Linux and have a terminal available, you can also try `netstat -tulpn` or `ss -tulpn` (depending on whats installed on your system). This will show you all listening ports. Port 3000 should be listed somewhere with state "LISTEN" and address "0.0.0.0".

![image](https://github.com/mikejac/node-red-contrib-google-smarthome/assets/2081806/f71ffc5a-4a55-491b-858d-c2482081ff5d)
(on my screenshot it's port 3001)

**Comment by Robi1010MD** [comment]
Hey Caprico,

maybe it is a problem with my setup? I’m using a windows server machine with a virtualization of home assistant in virtualbox and run node-red with the specific add-on.

I checked the netstat over cmd, unfortunately there was no entry.

Kind regards from germany.

Robert

**Comment by Caprico85** [comment]
Hey Robi1010MD,

Yes, I think it's related to your setup. ERR_CONNECTION_REFUSED means a port is not open or blocked. So somewhere in the chain of port forwardings between your router and node-red-contrib-google-smarthome some forwardings are not configured correctly.

On of your problems could be VirtualBox. A network package coming from your router reaches port 3001 on your PC. From there, Virtualbox needs to forward the package into your HomeAssistant container. So, in the settings of your virtualbox container, set up a port forwarding from port 3001 to a port (3001 or another port).

With this done, the port should now show up in netstat.

I don't know if anything needs to be done in Home Assistant too. It's been a while that I played around with HA. As far as I remember, some addons need additional configuration to expose their ports to the outside, while other addons have their ports exposed by default. I have no idea if Node-RED as HA-addon has it's ports exposed.

And finally, check if node-red-contrib-google-smarthome's webserver is started. Look on the debug panel. If the webserver is not started, there will be a message like "HTTP server is not running".

![image](https://github.com/mikejac/node-red-contrib-google-smarthome/assets/2081806/67755ccd-de85-42d4-86a6-d8bc36a8fc29)

Ebenfalls Grüße aus Deutschland :smile:

**Comment by Robi1010MD** [comment]
Ok,

still struggling. One of the bigger issues will be to create and upload an SSL Certificate on the fritzbox, instead of the the self-signed certificate. I think that the fritzbox certificate does not have the correct format (boxcert.cer) for using it.

I can fowards the ports within the fritzbox and virtualbox to my home assistant installation.

**Comment by Caprico85** [comment]
You don't need to upload the certificates on the Fritzbox. The Fritzbox just forwards all the traffic and doesn't decrypt it, therefore it doesn't need certificates. A certificate on the Fritzbox is only needed if you want to have your Fritbox admin interface available on the internet.

For certificate creation I usually recommend using a reverse proxy like Caddy, Traefik or Nginx Proxy Manager. Reverse proxies handle the certificate creation and renewal for you. Otherwise you would regularly have to recreate expired certificates. Personally, I use Caddy. My setup is a bit different though. For you, Nginx Proxy Manager might be a better option. Nginx Proxy Manager is available as HA addon. On of our users had written a few lines on how to use Nginx Proxy Manager [on the reverse proxy page](https://github.com/mikejac/node-red-contrib-google-smarthome/blob/master/docs/reverse_proxies.md).

**Comment by Robi1010MD** [comment]
Ok, what a ride. 

So i did this:

1. Router: Port Forwarding to my Windows Machine with virtualbox installation.
2. VirtualBox: Activate second network adapter, NAT Mode, add port forwarding rules from windows machine to home assistant vm with ports 3001.
3. Home Assistant: installed nginx proxy manager, followed the [instruction](https://github.com/mikejac/node-red-contrib-google-smarthome/blob/master/docs/reverse_proxies.md#nginx-proxy-manager-as-reverse-proxy) and got a ssl certificate for my specific domain.
4. Node Red: Somehow ssl offload doesn't work that way with nginx proxy manager!? anyway, pointed to installed certificates in my directory.
5. App: made to the login page from my google actions project. tried my login, result: error message could not reach [test] smarthome :'( 

Update:
It works now, had to point to the certificate in my directory. i don't know why it didn't work with ssl offload and the nginx proxy manager....

**Comment by Caprico85** [comment]
Glad you got it working. For the "could not reach" error, I would not have been of much help. I never had this problem myself and I have never been able to pinpoint the reason.
